Calico network policy (#2555)
projectcalico/calico: master/security/calico-network-policy.md

---
title: Get started with calico network policy
---

### Big Picture

> Enforce the network traffic that is allowed or denied using rules in Calico network policy. 

Reviewer: This sentence feels a bit awkward to me. Perhaps this is all it needs?
```suggestion
Enforce which network traffic is allowed or denied using rules in Calico network policy. 
```

Author: That is good catch. This doc has been through multiple reviewers and edits so I appreciate that you are doing a final look.

Author: Fixed.

    - Namespace selectors
    - Service account selectors
- **Optional packet handling controls**: disable connection tracking, apply before DNAT, apply to forwarded traffic and/or locally terminated traffic

### Concepts

#### Endpoints
Calico policies apply to **endpoints**. In Kubernetes, each pod is a Calico endpoint. However, Calico can support other kinds of endpoints. There are two types of Calico endpoints: **workload endpoints** (such as a Kubernetes pod or OpenStack VM) and **host endpoints** (an interface or group of interfaces on a host). 

#### Namespaced and global network policies
**Calico network policy** is a namespaced resource that applies to pods/containers/VMs in that namespace. 

<pre>
apiVersion: projectcalico.org/v3
kind: <b>NetworkPolicy</b>
metadata:
  <b>name</b>: allow-tcp-6379
  <b>namespace</b>: production
</pre>
{: .no-select-button}

**Calico global network policy** is a non-namespaced resource and can be applied to any kind of endpoint (pods, VMs, host interfaces) independent of namespace. 

<pre>
apiVersion: projectcalico.org/v3
kind: <b>GlobalNetworkPolicy</b>
metadata:
  <b>name</b>: allow-tcp-port-6379
</pre>
{: .no-select-button}

> **Calico network and global policies** are applied using calicoctl. Syntax is similar to Kubernetes, but there a few differences. For help, see [calicoctl user reference]({{site.baseurl}}/{{page.version}}/reference/calicoctl/).

Reviewer: The way this sentence is structured implies that it's **Calico global policy** but that's not a real thing!
```suggestion
**Calico network policies** and **Calico global network policies** are applied using calicoctl. Syntax is similar to Kubernetes, but there a few differences. For help, see [calicoctl user reference]({{site.baseurl}}/{{page.version}}/reference/calicoctl/).
```

Author: Yes, please change to: "Calico network policies and Calico global network policies are applied..."

Author: Fixed.

> |          Yes          |          No           |     Ingress     |
> |          No           |          Yes          |     Egress      |
> |          Yes          |          Yes          | Ingress, Egress |

#### Network traffic behaviors: deny and allow

The Kubernetes network policy specification defines the following behavior:
- If no network policies apply to a pod, then all traffic to/from that pod is allowed.
- If one or more network policies apply to a pod containing ingress rules, then only the ingress traffic specifically allowed by those policies is allowed, and all other traffic to/from the pod is denied.
- If one or more network policies apply to a pod containing egress rules, then only the egress traffic specifically allowed by those policies is allowed, and all other traffic to/from the pod is denied.

For compatibility with Kubernetes, **Calico network policy** follows the same behavior for Kubernetes pods.  For other endpoint types (VMs, host interfaces), Calico network policy is default deny. That is, only traffic specifically allowed by network policy is allowed, even if no network policies apply to the endpoint.

### Before you begin...

> None

Reviewer: I find these sections with `None` to be a bit jarring. Can we just remove the `Before you begin...` heading if there's nothing in it?

Author: This is Alex decision for the template. His reasoning is that when/if we ever get the ability to allow customers to comment inline, this invites "why is this blank, or I think there are requirements before you start."  I see your point, but for now, this is the decision. The how to template is a work in progress so if we get other inputs we will change it.
